### ISSUE TITLE
 Problems with 20150307 Gnome-Shell Theme

### ISSUE THREAD

**Issue by wolfland** [horst3180/Vertex-theme#58]
Horst3180,

First the good news,, The new Gnome-Shell theme looks awesome.

Now for the bad news..

First a little information. I am running Ubuntu Gnome 14.04(Gtk3.10/3.12), 14.10(Gtk3.14) and
15.04(Gtk3.14). I use the Vertex theme for both GDM and Gnome-Shell.

After I installed the 20150307 theme and either logged out or rebooted GDM would not start. I
reverted to the stock Gnome-Shell theme for GDM and was able to login in. I then set about
trying to find what was wrong as this configuration worked fine in the past.

Since the Gnome-Shell changes encompassed both the code and directory layout I took a 
dual approach and created 2 test versions. One had the new directory structure and I fixed
any obvious coding errors that I saw. On the second test version I fixed any obvious coding
errors as well as reverted to the original directory structure(e.g. everthing in one directory).

Some of the obvious coding issues were "url" statements with single quote(e.g. ') instead of
double quote(e.g. "), missing quotes, or incorrect path (for the test version with the new 
directory layout). Some of the areas where I found issues were the "page" selectors and
indicators, the "noise-texture.png" entry etc.

After a short time I had the two test versions completed and proceeded to see if I could get
the new theme working.

Symptoms: On Gtk3.10/Gtk3.12 no GDM login screen after installing the new Vertex 20150307
version, On Gtk3.14 the GDM login screen would appear but the first login would just hang and
you would always have to hit cancel and try logging in a second time which would work. The
theme seemed to work once logged in except for the things that were not working correctly
such as the application page selectors/indicators. 

I looked at the GDM logs and the only thing that made sense was the problem with the
"noise-texture.png" issue.

I tried using the changed test version with the new directory structure first and the GDM
login problem still persisted with the above symptoms. Of course the coding issues that
were addressed by my changes fixed the application page selectors etc. 

I then tried the changed code with the original all in one directory changes + the code
fixes and the GDM login screen returned to working as in previous Vertex versions.

Conculsions:

For some reason it appears that the previous "everything in one directory" file structure
is required to use Vertex as both the GDM and Gnome-Shell theme. I have looked at
over 10 other Gnome-Shell themes and they all use this same "everything in one
directory" file stucture so I suspect it may be mandatory.

Most of the coding issues were single quotes instead of double quotes or missing
quotes from "url" statements. I only tried to resolve the obvious things that could be
done quickly and so may have missed other coding ssues.

So in short I had to revert back to the "all in one directory" file structure to get things
back working again as well as correct some minor coding issues to get every thing 
working.

Let me know if I missed something that would allow you to keep the new file stucture
and allow using the theme for both GDM and Gnome-Shell.

I will be glad to test any changes if needed.

There does not seem to be anyway to attach my code changes to this problem so
I will send them in an email.

Oh almost forgot.. While I am here there are some minor problems with flashback(metacity)
in 14.04(Gtk3.10) and 15.04(Gtk3.14). In 14.04 there is a problem with the "Application and 
Places" buttons on the top panel which I have a fix for. In both 14.04 and 15.04 there are
problems with the bottom panel being partially white (buttons and window switcher) with the
Vertex theme. The Vertex-Dark theme works fine in both 14.04 and 15.04. I will attach screen
shot here:

![screenshot from 2015-03-02 21_52_36](https://cloud.githubusercontent.com/assets/8964224/6544867/99b6f450-c532-11e4-9551-23572ec91b8d.png)
![screenshot from 2015-03-02 21_53_01](https://cloud.githubusercontent.com/assets/8964224/6544870/a5f9cb98-c532-11e4-8315-3fdc6c1ea764.png)

I will attach a version of gnome-applications.css to my email that has my change for 14.04
flashback "Applications and Places" buttons in it.

Thanks again for your awesome themes and keep up the good work,
WolfLand

P.S. I will have to look into Sass as I have my 14,04 and 14,10 machines set up to compile
gtk themes but I have not had tine to work on it at the moment.

**Comment by horst3180** [comment]
Hi, thanks for reporting. Did you install the theme to /usr/share/gnome-shell/theme? I haven't tested this method at all and I don't think it's a good idea, because it can cause unexpected behaviour and faulty themes lead to gdm crashes (as you have experienced).

However there are (were) some wrong image url's in the theme (should have checked before releasing it). Normally this wouldn't be a huge deal because gnome-shell would either fall back to the default theme or just not display the affected images.
I fixed the wrong url's (see d496f80ad86dc04c829ba8a2f990afe2dcb16b17) and installed the theme to /usr/share/gnome-shell/theme for a test. Gdm doesn't crash, but I can confirm the login issue. Additionally if I lock the screen I wont be able to unlock it and I have to restart gdm. Did you fix this issue?

The quotes in the url statements are not errors, actually. Css supports double, single or no quotes if the url doesn't have special characters, but it would look nicer if I made this consistent.

I don't think the directory structure is mandatory. Look at this theme 
http://www.deviantart.com/art/Ciliora-Secunda-Shell-495066314
It places all images in subdirectories too, so I don't think that's a problem.

About flashback. I though it was removed some time ago (around 3.10 maybe)? Well, but if it still exists, I'll have a look at it.

So, long story short:
The only remaining problem right now is not being unable to login the first time and unlock the screen. If I can make this work properly I may think about styling the rest of the login/lock screen and make it look better. And I will try to fix the flashback mode.

**Comment by wolfland** [comment]
horst3180,

I used the following method which uses a soft link instead of actually 
installing the Vertex theme
to the /usr/share/gnome-shell/theme directory. This ensures an easy way 
to change back if there
is a problem. I guess it really doesn't matter which method is used so 
long as the original theme
is backed up:

This comes from the NovaShell theme on gnome-look.org web site.

Installation as default theme: (NovaShell will be the !default! theme 
for all users and gdm):
*copy NovaShell-directory into /usr/share/themes/[/li]
*rename /usr/share/gnome-shell/theme to 
/usr/share/gnome-shell/theme.original
$ sudo mv /usr/share/gnome-shell/theme /usr/share/gnome-shell/theme.original
*create a symbolic link to /usr/share/gnome-shell/theme
$ sudo ln -s /usr/share/themes/NovaShell/gnome-shell 
/usr/share/gnome-shell/theme
*log out and log in again

Thanks for the info about the quotes. I don't have much experience with 
Gnome Shell
themes yet and wanted to eliminate them as causing the problem.

On 03/08/2015 08:07 AM, horst3180 wrote:
>
> Hi, thanks for reporting. Did you install the theme to 
> /usr/share/gnome-shell/theme? I haven't tested this method at all and 
> I don't think it's a good idea, because it can cause unexpected 
> behaviour and faulty themes lead to gdm crashes (as you have experienced).
>
> However there are (were) some wrong image url's in the theme (should 
> have checked before releasing it). Normally this wouldn't be a huge 
> deal because gnome-shell would either fall back to the default theme 
> or just not display the affected images.
> I fixed the wrong url's (see d496f80 
> <https://github.com/horst3180/Vertex-theme/commit/d496f80ad86dc04c829ba8a2f990afe2dcb16b17>) 
> and installed the theme to /usr/share/gnome-shell/theme for a test. 
> Gdm doesn't crash, but I can confirm the login issue. Additionally if 
> I lock the screen I wont be able to unlock it and I have to restart 
> gdm. Did you fix this issue?
>
The only way I was able to fix the above issue was to go back to the 
original "everything in one directory" file structure. This error was 
different in Gtk3.10/Gtk3.12 and Gtk3.14. Gtk3.10/Gtk3.12
failed to give the GDM login screen at all. Gtk3.14 was able to give the 
login screen but with the
login problem you note above.
>
> The quotes in the url statements are not errors, actually. Css 
> supports double, single or no quotes if the url doesn't have special 
> characters, but it would look nicer if I made this consistent.
>
> I don't think the directory structure is mandatory. Look at this theme
> http://www.deviantart.com/art/Ciliora-Secunda-Shell-495066314
> It places all images in subdirectories too, so I don't think that's a 
> problem.
>
It may be that one of the other typo's caused the problems so I will 
pull from the git this morning
and try it again with the corrected code and new directory structure 
again. Hopefully it will work
or at least I can capture the GDM logs (not sure how much they will 
help) or something if it fails.

The file stucture should not make a difference so maybe there is 
something else wrong that I
missed. I will grab the Ciliora-Secunda-Shell theme and look at it and 
do a stare and compare
as well.
>
> About flashback. I though it was removed some time ago (around 3.10 
> maybe)? Well, but if it still exists, I'll have a look at it.
>
I thought it went away in Ubuntu 14.10 (Gtk3.12) but it is definitely 
back in 15.04(Gtk3.14) as I have
it in both my Elite (Unity+GS+Gnome Classic+Flashback metacity and 
compiz+CairoDock+XBMC) as
well as my 32bit Gnome Shell version (for my old HP laptop). So I have 
it running in both 14.04 and
15.04.
>
> So, long story short:
> The only remaining problem right now is not being unable to login the 
> first time and unlock the screen. If I can make this work properly I 
> may think about styling the rest of the login/lock screen and make it 
> look better. And I will try to fix the flashback mode.
>
> —
>
Thanks. I will retest with the latest git version and grab the GDM logs 
if it still fails. I will also do a
stare and compare with  Ciliora-Secunda-Shell and see if I see any thing 
obvious. If we still get problems I will install and try it as well to 
see what happens.

Thanks again and I will reply with results as soon as I have them.

WolfLand

**Comment by horst3180** [comment]
I edited my previous reply, it seems you didn't get notified via email. Turns out you were right and I will push some fixes today.

**Comment by wolfland** [comment]
horst3180,

                 I just got through testing and the default gnome shell 
theme/gdm theme has to be
                 all in the same directory. The only drawback I could 
see to what you were doing
                 was that you would need a default/gdm login theme as 
well as the user theme
                 (e.g. both directory formats) which would cause twice 
the amount of work any
                 time you wanted to make changes.

                 I think this is why most Gnome Shell themes are in the 
"all in one directory" format.

                 Of course the Gnome folks are always changing things so 
it will be interesting to see
                 what happens to Gnome Shell themes when compiling 
becomes a requirement.

Thanks,
WolfLand

**Comment by horst3180** [comment]
I will change the directory structure back to the old one (in fact I already did this in a local branch). I still have to work out some bugs with the login screen etc. but I think it should work.

For Gnome 3.16 the theme has to be compiled into a .gresource file. It should be possible to do that with a script, but I'm not sure if it's worth the effort just for the login screen.

**Comment by horst3180** [comment]
Oh, and I had a look at the flashback mode. I fixed it up a bit, but it turns out the black parts of the panel in your screenshot are a bit buggy. Setting a gradient/box-shadow on them doesn't work as it should, so I can't style them properly. I think this is a bug that should be reported somewhere.

**Comment by wolfland** [comment]
Great.. The login screen was working fine before the 20150307 update. 
The only thing I noticed was
             the password box having a dark background instead of the 
normal light background. It seems
             to be working fine for me since I changed the directory 
structure back. I verified that the lock
             screen is working properly and the login works first time.

             I just added the dev files to my 15.04 Elite respin and 
have already added them to 14.04 and
             14.10 as well as my 32bit 15.04 GnomeShell respin. One of 
these days I will get around to
             trying them out and building a gresources file.

             I agree. It seems like a lot of extra effort as well as 
having to update 2 copies of the same code
             all of the time. The Gnome folks should have a better 
though out system. There is no reason
             why one theme cannot be used for both the default/gdm and 
user themes. There also needs
             to be an easier way to theme gdm that we currently have. We 
should be able to completely
             control all gdm theme attribuTes(e.g. background, theme, 
icons, cursor etc) from system
             settings instead of having to jump through hoops just to 
change the theme only.

             Thanks again for your hard work and great themes.

WolfLand

**Comment by wolfland** [comment]
Yeah  I noticed the black too and it was not like that when I built the 
alpha version so something
has definitely changed. There is also a metacity issue that I have 
seeing intermittently.  It seems
like I may have had a similar problem with gradient/box shadows when I 
tried to fix the bottom
panel myself in 14.04 flashback but it only seemed to be the bottom 
panel that had the problem
if I remember correctly.

I will check and see if these issues have been reported already and if 
not I will submit bug reports
for them early next week and let you know the number so that you can 
track their progress(also
if more than 1 person sees the problem then they are more likely to fix 
them).

Thanks again,
WolfLand

**Comment by wolfland** [comment]
Horst3180,

                  Thanks for the flashback changes they are certainly 
much better than before.

                  I will open a bug report when beta happens on/after 
the 26th if it doesn't slip.

                  I know you are busy trying to get ready for the 
release of Gnome 3.16 this month
                  so please put the login theme issue on the back burner 
if you need to. Any solution
                  (including none) will be ok as I can always write a 
bash/awk/sed script to convert the
                  user style into the login style format.

                  I am not up on sass but if it would allow you to just 
write one code set and then output
                  it in both formats (e.g. user and default/login) that 
would be one big reason to justify it's
                  use. This would prevent having to upkeep 2 sets of 
code for the same thing. I guess a
                  bash/awk/sed script would achieve the same goal but 
you would still have to upkeep the
                  code and script.

                  I also noticed you had an item to add a plank theme to 
your theme. I have been using
                  the "Wingy" theme which kinda looks like the panel in 
Vertex Gnome Shell. I didn't
                  look at the item so I did not see if the request was 
for a "panel" style theme or OSX
                  style theme or one more like the default plank theme. 
I took a quick look at plank
                  theming and it looks pretty straight forward. I make a 
theme to match with the
                  budgie desktop panel on my laptop and it only took a 
few minutes.

                  I will attach a file with some plank themes for you to 
try.

                  Thanks again for all your hard work. Gnome 3.16 is 
going to be a real challenge and
                  I am hoping that someone creates a Ubuntu 15.04 PPA 
like they did for 3.12 and 3.14
                  so I can upgrade to 3.16 after it and Ubuntu 15.04 
Gnome releases.

Thanks,
WolfLand

**Comment by horst3180** [comment]
Actually I can also use the all-in-one directory structure when the theme is installed normally, too, so that would be no problem. In other words: If the theme works correctly as login theme it will work correctly as user theme. The only sacrifice would be a messy gnome-shell theme directory with all images scattered around.
I have seen some other small bugs when using the theme as login theme, that I want to fix. Let me push my local branch with some fixes real quick, then you can test it, if you want. When I'm done fixing I'll merge this into master.
The gnome-shell 3.16 theme will need those fixes too, but I can't merge them because it's ported to Sass, so the whole thing might take a while.

**Comment by horst3180** [comment]
Ok, check out the wip/gnome-shell-login branch if you want to.
Also the 3.16 version of the theme is mostly finished. I'm just waiting for the gnome hard code freeze on monday to push it into master.

**Comment by wolfland** [comment]
Thanks, no need to hurry on my behalf as I am using the version that I 
tested with which
has your new directory structure for the user theme and my test code 
"all in one directory"
structure for the login style for now. I will give your changes a test 
as soon as I can. I am
not sure how long it will take for someone to make a Gnome 3.16 PPA for 
Ubuntu but typically
it seems to take at least a month after the version is released. I may 
have to grab the code
out of git and compile it if I want Gnome 3.16 any earlier.

Thanks again,
Brett

**Comment by horst3180** [comment]
Yeah, compiling Gnome is fun. I updated my JHBuild Gnome-git install today and it took several hours on my laptop. I have to say that I'm pretty pleased with 3.16, so far. I hope that it will hit the Arch repos quickly after release.

**Comment by wolfland** [comment]
I completed testing the theme changes last night and it works fine and 
looks great. I would have
got done sooner but had to switch a hard drive out on a laptop to get 
Gtk3.10.

I tested Gtk3.10, Gtk3.12 and Gtk3.14 without any issues.

I checked out Gtk3.16 on ISO yesterday and found a Ubuntu Gnome 3.16 PPA 
too but it is not
fully populated yet and has'nt been built for about 5 weeks. I suspect 
they will start populating
and rebuilding the .deb files after gnome 3.16 releases on the 25th. If 
not I will compile it from
jhbuild files once I move my 15.04's to their perspective build machines 
upon beta.

I really liked the changes for Gnome 3.16 so I am going to use it for my 
15.04 Gnome version
and will compile it from jhbuild if I have to until the PPA becomes ready.

Thanks for your hard work and I can't wait to try out your 3.16 version 
when I get it running.

Thanks Again,
WolfLand

**Comment by horst3180** [comment]
I'll close this. The remaining issue is a duplicate of https://github.com/horst3180/Vertex-theme/issues/2

**Comment by wolfland** [comment]
Horst3180,

                  I finally got a chance to bug report the black panel 
background problem.
                  The bug report number is: 1439871.

                  I also reported another problem with artifacts on 
metacity windows with Vertex.
                  This one is not theme related as a "metacity --replace 
&" (e.g. metacity restart)
                  fixes the problem.
                  The bug report number is: 1439860

Thanks,
WolfLand
